lib/elixir/src/elixir_erl.erl

```diff
@@ -373,12 +373,12 @@ specs_form(Map, Data, Defmacro, Unreachable, Forms) ->
      {Kind, Expr, Caller} <- take_type_spec(Data, macrocallback)],
 
   Optional = lists:flatten(take_type_spec(Data, optional_callbacks)),
-  SpecsForms = specs_form(spec, Specs, Unreachable, [], Defmacro, Forms),
+  SpecsForms = specs_form(spec, Specs, Unreachable, [], Defmacro, Forms, Map),
```

Comment:
Unreachable is also available in the `Map` and I think the only reason we pass it all the way down here is because we didn't pass the `Map`. But now that we do, I would change it so no arity of `specs_form` receive Unreachable and we also change `dynamic_form` to no longer return it.